File: pact-jvm-provider-junit/src/main/java/au/com/dius/pact/provider/junit/sysprops/SystemPropertyResolver.java
package au.com.dius.pact.provider.junit.sysprops;

import org.apache.commons.lang3.StringUtils;

public class SystemPropertyResolver implements ValueResolver {

  @Override
  public String resolveValue(final String property) {
    PropertyValueTuple tuple = new PropertyValueTuple(property).invoke();
    String propertyValue = System.getProperty(tuple.getPropertyName());
    if (propertyValue == null) {
      propertyValue = System.getenv(tuple.getPropertyName());
    }
    if (propertyValue == null) {
      propertyValue = tuple.getDefaultValue();
    }
    if (propertyValue == null) {
      throw new RuntimeException("Could not resolve property \"" + tuple.getPropertyName()
        + "\" in the system properties or environment variables and no default value is supplied");
    }
    return propertyValue;
  }

  @Override
  public boolean propertyDefined(String property) {
    String propertyValue = System.getProperty(property);
    if (propertyValue == null) {
      propertyValue = System.getenv(property);
    }
    return propertyValue != null;
  }

  private class PropertyValueTuple {
    private String propertyName;
    private String defaultValue;

    PropertyValueTuple(String property) {
      this.propertyName = property;
      this.defaultValue = null;
    }

    String getPropertyName() {
      return propertyName;
    }

    String getDefaultValue() {
      return defaultValue;
    }

    PropertyValueTuple invoke() {
      if (propertyName.contains(":")) {
        String[] kv = StringUtils.splitPreserveAllTokens(propertyName, ':');
        propertyName = kv[0];
        if (kv.length > 1) {
          defaultValue = kv[1];
        }
      }
      return this;
    }
  }
}

File: pact-jvm-support/src/main/kotlin/au/com/dius/pact/support/expressions/SystemPropertyResolver.kt
package au.com.dius.pact.support.expressions

import org.apache.commons.lang3.StringUtils

class SystemPropertyResolver : ValueResolver {

  override fun resolveValue(property: String): String {
    val tuple = PropertyValueTuple(property).invoke()
    var propertyValue: String? = System.getProperty(tuple.propertyName!!)
    if (propertyValue == null) {
      propertyValue = System.getenv(tuple.propertyName)
    }
    if (propertyValue == null) {
      propertyValue = tuple.defaultValue
    }
    if (propertyValue == null) {
      throw RuntimeException("Could not resolve property \"${tuple.propertyName}\" in the system properties or " +
        "environment variables and no default value is supplied")
    }
    return propertyValue
  }

  override fun propertyDefined(property: String): Boolean {
    var propertyValue: String? = System.getProperty(property)
    if (propertyValue == null) {
      propertyValue = System.getenv(property)
    }
    return propertyValue != null
  }

  class PropertyValueTuple(property: String) {
    var propertyName: String? = null
      private set
    var defaultValue: String? = null
      private set

    init {
      this.propertyName = property
      this.defaultValue = null
    }

    operator fun invoke(): PropertyValueTuple {
      if (propertyName!!.contains(":")) {
        val kv = StringUtils.splitPreserveAllTokens(propertyName, ':')
        propertyName = kv[0]
        if (kv.size > 1) {
          defaultValue = kv[1]
        }
      }
      return this
    }
  }
}
